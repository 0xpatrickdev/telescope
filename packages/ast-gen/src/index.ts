export * from './types';
export * from './amino';
export * from './client';
export * from './messages';
export * from './utils';