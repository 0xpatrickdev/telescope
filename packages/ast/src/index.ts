export * from './encoding';
export * from './clients';
export * from './registry';
export * from './services';
export * from './utils';
export * from './bundle';
export * from './types';