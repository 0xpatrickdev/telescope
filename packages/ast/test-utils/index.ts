import generate from '@babel/generator';
import { ProtoStore, getNestedProto, parseProto } from '@osmonauts/proto-parser'
import { ProtoType } from '../src/encoding/proto/types';
import { ParseContext } from '../src/encoding/context';

export const expectCode = (ast) => {
    expect(
        generate(ast).code
    ).toMatchSnapshot();
}
export const printCode = (ast) => {
    console.log(
        generate(ast).code
    );
}

export const prepareContext = (store: ProtoStore, protoFile: string) => {
    const ref = store.findProto(protoFile);
    const traversed = ref.traversed
    if (!traversed) {
        throw new Error('not traversed!');
    }
    const proto: Record<string, ProtoType> = getNestedProto(traversed);
    const protos: ProtoType[] = Object.values(proto).filter(
        proto => proto.name.startsWith('Msg')
            &&
            !proto.name.endsWith('Response')
            &&
            proto.name !== 'Msg'
    );

    const context = new ParseContext(ref, store);

    return {
        context,
        root: traversed,
        protos
    };
};
