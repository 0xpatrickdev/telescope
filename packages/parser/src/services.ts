import { ProtoField } from '@osmonauts/types';

interface ServiceOptions {
    "(google.api.http).get": string;
}
export const parseServiceUrl = (
    options: ServiceOptions
) => {
    const url = options?.['(google.api.http).get'];
    if (!url) return;
    const match = url.match(/\{([^\}]*)\}/g);
    // NOTE pathParams are the original casing!
    // e.g.: 
    // "(google.api.http).get": "/cosmwasm/wasm/v1/code/{code_id}",
    // "(google.api.http).get": "/osmosis/gamm/v1beta1/pools/{poolId}/total_pool_liquidity",
    return {
        url,
        pathParams: match?.length ? match.map(el => el.replace('{', '').replace('}', '')) : []
    };
};

// this basically takes in a ServiceMethod object! not any!
export const parseService = (
    obj: any
) => {
    const fields: Record<string, ProtoField> = obj.fields ?? {};
    const options: ServiceOptions = obj.options ?? {};
    const pathInfo = parseServiceUrl(options);
    if (!pathInfo) return;
    const allParams = Object.keys(fields).map(field => {
        // NOTE pathParams are the original casing!
        return fields[field].options['(telescope:orig)'];
    })
    const queryParams = allParams.filter(param => !pathInfo.pathParams.includes(param));
    return {
        ...pathInfo,
        queryParams,
        casing: {}
    };
};

// ProtoField:
// validatorAddress: {
//     parsedType: { name: 'string', type: 'native' },
//     isScalar: true,
//     typeNum: 13,
//     type: 'string',
//     id: 1,
//     options: {
//       '(cosmos_proto.scalar)': 'cosmos.AddressString',
//       '(telescope:name)': 'validatorAddress',
//       '(telescope:orig)': 'validator_address',
//       '(telescope:camel)': 'validatorAddress'
//     },
//     comment: 'validator_address defines the validator address to query for.'
//   }