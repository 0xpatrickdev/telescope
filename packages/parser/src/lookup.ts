import { traverse } from '@babel/core';
import { ProtoStore } from './store';
import { ProtoRef, ProtoRoot } from './types';
import { getNestedProto } from './utils';

export interface Lookup {
    obj: any;
    name: string;
    import: string;
    importType: string;
    importedName: string;
    package: string;
}

export const recursiveLookup = (proto: any, name: string, scope: string[] = [], allowNested = true) => {
    if (!proto) return;

    if (proto.hasOwnProperty(name)) return {
        name,
        scope,
        ...proto[name]
    }

    if (allowNested && proto) {
        const keys = Object.keys(proto);
        for (let k = 0; k < keys.length; k++) {
            const found = recursiveLookup(proto[keys[k]].nested, name, [...scope, keys[k]], allowNested);
            if (found) {
                return {
                    name,
                    ...found
                };
            };
        }
    }
};

const getRoot = (ref: ProtoRef) => {
    if (ref.traversed) return ref.traversed;
    return ref.proto;
};

/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "Empty"

  finds Empty inside of the import

*/
export const importLookup = (store: ProtoStore, ref: ProtoRef, name: string) => {
    const root = getRoot(ref);
    const objectsFromImports = root?.imports?.map(imp => {
        const ref = store.findProto(imp);
        return {
            name,
            importType: 'import',
            import: imp,
            importedName: name,
            obj: lookup(store, ref, name, false),
        };
    }).filter(a => !!a.obj) ?? []
    if (objectsFromImports.length) return objectsFromImports[0];
};


/*

  "imports": [
    "google/api/expr/v1alpha1/syntax.proto",
    "google/protobuf/empty.proto",
    "google/protobuf/struct.proto"
  ],

  ...

  "type": "google.protobuf.NullValue"

  finds NullValue directly from package name

*/

export const protoImportLookup = (
    store: ProtoStore,
    ref: ProtoRef,
    name: string
): Lookup => {
    const root = getRoot(ref);
    if (name.startsWith('.')) name = name.replace(/^\./, '');
    const nameAsArray = name.split('.');
    const objectName = nameAsArray.pop();
    const packageName = nameAsArray.join('.')
    const objs = store.getProtos()
        // 1 ask all package files of package, e.g "google.protobuf"
        .filter(proto => proto.proto.package === packageName)
        // 2 cross-checking w the imports by filter()
        .filter(proto => root.imports.includes(proto.filename))
        .map((ref: ProtoRef) => {
            return {
                import: ref.filename,
                obj: lookup(store, ref, objectName, false)
            }
        })
        .filter(a => !!a.obj);
    if (objs.length) {
        return {
            import: objs[0].import,
            importType: 'protoImport',
            obj: objs[0].obj,
            importedName: name,
            name: objectName,
            package: packageName
        }
    }
};

export const lookup = (
    store: ProtoStore,
    ref: ProtoRef,
    name: string,
    allowNested = true
) => {
    const root = getRoot(ref);
    const nested = getNestedProto(root);
    return recursiveLookup(nested, name, [root.package], allowNested);
};

export const lookupAny = (
    store: ProtoStore,
    ref: ProtoRef,
    name: string
): Lookup => {
    let refObject = lookupLocal(store, ref, name);

    if (refObject) {
        return refObject;
    }

    refObject = externalLookup(store, ref, name);
    if (refObject) {
        return refObject;
    }

};

export const lookupLocal = (
    store: ProtoStore,
    ref: ProtoRef,
    name: string
): Lookup => {
    const root = getRoot(ref);
    let refObject = lookup(store, ref, name);

    if (refObject) {
        return {
            name,
            import: ref.filename,
            importType: 'local',
            importedName: name,
            package: root.package,
            obj: refObject
        }
    }
};

export const externalLookup = (
    store: ProtoStore,
    ref: ProtoRef,
    name: string
): Lookup => {
    let refObject;

    refObject = protoImportLookup(store, ref, name);
    if (refObject) {
        return refObject;
    }

    refObject = importLookup(store, ref, name);
    if (refObject) {
        return refObject;
    }

};
