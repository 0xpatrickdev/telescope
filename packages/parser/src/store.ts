import { sync as glob } from 'glob';
import { parse } from '@pyramation/protobufjs';
import { readFileSync } from 'fs';
import { join, resolve as pathResolve } from 'path';
import { ProtoDep, ProtoRef, ProtoServiceMethod, TelescopeOptions } from '@osmonauts/types';
import { getNestedProto, getPackageAndNestedFromStr } from './utils';
import { traverse } from './traverse';
import { lookupAny, lookupAnyFromImports } from './lookup';
import { defaultTelescopeOptions } from '@osmonauts/types';

import google_any from './native/any';
import google_duration from './native/duration';
import google_timestamp from './native/timestamp';
import google_empty from './native/empty';
import google_field_mask from './native/field_mask';
import google_struct from './native/struct';
import google_wrappers from './native/wrappers';

const GOOGLE_PROTOS = [
    ['google/protobuf/any.proto', google_any],
    ['google/protobuf/duration.proto', google_duration],
    ['google/protobuf/timestamp.proto', google_timestamp],
    ['google/protobuf/empty.proto', google_empty],
    ['google/protobuf/struct.proto', google_struct],
    ['google/protobuf/wrappers.proto', google_wrappers],
    ['google/protobuf/field_mask.proto', google_field_mask]
];

interface ParseProtoOptions {
    keepCase?: boolean;
    alternateCommentMode?: boolean;
    preferTrailingComment?: boolean;
}

const protoParseOptionsDefaults = {
    keepCase: false,
    alternateCommentMode: true,
    preferTrailingComment: false
};

export const parseProto = (content, options?: ParseProtoOptions) => {
    if (!options) {
        options = protoParseOptionsDefaults
    }
    return parse(content, options);
};
export class ProtoStore {
    files: string[];
    protoDirs: string[];
    deps: ProtoDep[];
    protos: ProtoRef[];
    packages: string[];
    options: TelescopeOptions;

    requests: Record<string, ProtoServiceMethod> = {};

    _traversed: boolean = false;

    constructor(protoDirs: string[] = [], options: TelescopeOptions = defaultTelescopeOptions) {
        this.protoDirs = protoDirs.map(protoDir => pathResolve(protoDir));
        this.options = options;
    }

    findProto(filename): ProtoRef {
        return this.getProtos().find(proto => {
            return proto.filename === filename
        });
    }

    findProtoWhere(fn: (ref: ProtoRef) => boolean): ProtoRef {
        return this.getProtos().find(ref => {
            return fn(ref)
        });
    }

    filterProtoWhere(fn: (ref: ProtoRef) => boolean): ProtoRef[] {
        return this.getProtos().filter(ref => {
            return fn(ref)
        });
    }

    findProtoObject(filename, name): any {
        const proto = this.findProto(filename);
        return getNestedProto(proto.traversed ?? proto.proto)[name];
    }

    registerRequest(requestType: string, svc: ProtoServiceMethod): void {
        this.requests[requestType] = svc;
    };

    getProtos(): ProtoRef[] {
        if (this.protos) return this.protos;
        const contents = this.protoDirs.reduce((m, protoDir) => {
            const protoSplat = join(protoDir, '/**/*.proto');
            const protoFiles = glob(protoSplat);
            const contents = protoFiles.map(filename => ({
                absolute: filename,
                filename: filename.split(protoDir)[1].replace(/^\//, ''),
                content: readFileSync(filename, 'utf-8')
            }));
            return [...m, ...contents];
        }, []);

        // if they don't got it, let's give it to 'em!
        GOOGLE_PROTOS.map(([f, v]) => {
            const found = contents.find(file => file.filename === f);
            if (!found) {
                contents.push({
                    absolute: f,
                    filename: f,
                    content: v
                });
            }
        });

        const protos = contents.map(({ absolute, filename, content }) => {
            try {
                const proto = parseProto(content, this.options.prototypes.parser);
                return {
                    absolute,
                    filename,
                    proto,
                };
            } catch (e) {
                console.error(`${filename} has a proto syntax error`)
                throw e;
            }
        });

        this.protos = protos;
        return protos;
    }

    getPackages(): string[] {
        if (this.packages) return this.packages;
        this.packages = this.getProtos().reduce((m, ref) => {
            return [...new Set([...m, ref.proto.package])];
        }, []);
        // LONGEST strings first, for better matching
        this.packages = this.packages.sort((a, b) => {
            return b.length - a.length;
        })
        return this.packages;
    }

    parseScope(type: string) {
        const pkgs = this.getPackages();
        for (let pkg of pkgs) {
            const found = getPackageAndNestedFromStr(type, pkg);
            if (found) return found;
        }
    }

    getDeps(): ProtoDep[] {
        if (this.deps) return this.deps;
        const deps = this.getProtos().map(el => {
            const {
                filename,
                proto: {
                    package: pkg,
                    imports
                }
            } = el;
            return {
                filename,
                package: pkg,
                imports
            }
        });
        this.deps = deps;
        return deps;
    }

    traverseAll(): void {
        if (this._traversed) return;
        this.protos = this.getProtos().map((ref: ProtoRef) => {
            return {
                absolute: ref.absolute,
                filename: ref.filename,
                proto: ref.proto,
                traversed: traverse(this, ref)
            };
        })
        this._traversed = true;
    }

    get(from: ProtoRef, name: string) {
        if (!this._traversed) throw new Error('get() requires traversal')
        return lookupAny(this, from, name);
    }

    getImportFromRef(ref: ProtoRef, name: string) {
        if (!this._traversed) throw new Error('getImportFromRef() requires traversal')
        return lookupAnyFromImports(this, ref, name);
    }

}
