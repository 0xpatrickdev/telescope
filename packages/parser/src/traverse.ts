import { Service, Type, Field, Enum, Root, Namespace } from 'protobufjs';
import { importLookup, protoImportLookup, lookup, lookupAny } from './lookup';
import { ProtoStore } from './store';
import { ProtoRoot, ProtoRef } from './types';
import { instanceType, NATIVE_TYPES } from './utils';

export const traverse = (store: ProtoStore, ref: ProtoRef) => {
    const imports = {};
    const exports = {};

    const obj: ProtoRoot & {
        parsedImports: Record<string, string[]>;
        parsedExports: Record<string, any>;
        importNames: Record<string, any>;
    } = {
        imports: ref.proto.imports,
        package: ref.proto.package,
        root: recursiveTraversal(
            store,
            ref,
            ref.proto.root,
            imports,
            exports,
            false
        ),
        parsedImports: null,
        parsedExports: null,
        importNames: null,
    };

    obj.parsedImports = imports;
    obj.parsedExports = exports;
    let counter = 1;
    obj.importNames = Object.entries(imports).reduce((m, [path, names]) => {
        m[path] = m[path] || {};
        names.forEach(importName => {
            const hasConflict = Object.entries(imports).some(([otherPath, otherNames]) => {
                if (path === otherPath) return false;
                if (otherNames.includes(importName)) return true;
            });
            if (hasConflict || exports.hasOwnProperty(importName)) {
                m[path][importName] = importName + counter++;
            } else {
                m[path][importName] = importName;
            }
        })
        return m;
    }, {});
    return obj;
};

const traverseFields = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object) => {
    return Object.keys(obj.fields).reduce((m, key) => {

        const field = obj.fields[key];

        const serialize = () => {
            if (typeof field.toJSON !== 'undefined') {
                // non-traversed
                return field.toJSON({ keepComments: true });
            }
            // traversed
            return field;
        }

        let found: any = null;

        if (NATIVE_TYPES.includes(field.type)) {
            m[key] = {
                parsedType: { name: field.type, type: 'native' },
                isScalar: true,
                typeNum: NATIVE_TYPES.indexOf(field.type),
                ...serialize()
            }
            return m;
        }

        // local scope first
        found = lookup(store, ref, field.type);
        if (found) {
            m[key] = {
                scope: found.scope,
                parsedType: instanceType(found),
                ...serialize(),
            };
            return m;
        }

        found = importLookup(store, ref, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {
                parsedType: instanceType(found.obj),
                scopeType: 'import',
                scope: [found.obj.scope],
                ...serialize(),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        found = protoImportLookup(store, ref, field.type);
        if (found) {
            imports[found.import] = imports[found.import] || [];
            imports[found.import] = [...new Set([...imports[found.import], found.name])];
            m[key] = {

                parsedType: instanceType(found.obj),
                scopeType: 'protoImport',
                scope: [found.package],
                ...serialize(),
                importedName: found.importedName,
                import: found.import,
            };
            return m;
        }

        return m;
    }, {});
};

const traverseType = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object, exports: object, isNested: boolean) => {
    let nested = null;
    if (obj.nested) {
        nested = Object.keys(obj.nested).reduce((m, key) => {
            m[key] = recursiveTraversal(store, ref, obj.nested[key], imports, exports, true);
            return m;
        }, {});
    }

    if (!isNested) {
        exports[obj.name] = exports[obj.name] || [];
        exports[obj.name].push({
            name: obj.name
        })

    }

    const traversed = {
        type: 'Type',
        name: obj.name,
        options: obj.options,
        oneofs: obj.oneofs ? Object.keys(obj.oneofs).reduce((m, v) => {
            m[v] = {
                // parse oneof
                oneof: obj.oneofs[v].oneof.map(name => name)
            };
            return m;
        }, {}) : undefined,
        fields: traverseFields(store, ref, obj, imports),
        nested,
        keyTypes: []
    };

    // parse keyType
    const hasKeyType = Object.keys(traversed.fields).some(field => !!traversed.fields[field].keyType);
    let keyTypes = [];
    if (hasKeyType) {
        keyTypes = Object.keys(traversed.fields)
            .filter(field => !!traversed.fields[field].keyType)
            .map(field => {
                return {
                    name: field,
                    ...traversed.fields[field]
                };
            });
    }

    traversed.keyTypes = keyTypes;
    return traversed;
};

const traverseEnum = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object) => {
    return {
        type: 'Enum',
        name: obj.name,
        ...obj.toJSON({ keepComments: true })
    }
};

const traverseField = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object) => {
    return {
        type: 'Field',
        name: obj.name,
        ...obj.toJSON({ keepComments: true })
    }
};

const traverseServiceMethod = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object, name: string) => {

    const service = obj.methods[name];
    const { requestType } = service;
    let refObject = lookupAny(store, ref, requestType);

    if (!refObject) {
        throw new Error('Symbol not found ' + requestType);
    }

    return {
        type: 'ServiceMethod',
        name,
        requestType,
        fields: traverseFields(store, ref, refObject.obj, imports)
    };
};

const getServiceType = (obj) => {
    if (obj.name === 'Msg') return 'Mutation';
    if (obj.name === 'Query') return 'Query';
    return 'Unknown';
}

const traverseService = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object) => {
    const methods = Object.keys(obj.methods).reduce((m, key) => {
        m[key] = traverseServiceMethod(
            store, ref, obj, imports, key
        );
        return m;
    }, {})

    return {
        type: 'Service',
        name: obj.name,
        serviceType: getServiceType(obj),
        ...obj.toJSON({ keepComments: true }),
        parsedMethods: methods
    }
};

export const recursiveTraversal = (store: ProtoStore, ref: ProtoRef, obj: any, imports: object, exports: object, isNested: boolean) => {
    if (obj instanceof Type) {
        return traverseType(store, ref, obj, imports, exports, isNested);
    }
    if (obj instanceof Enum) {
        return traverseEnum(store, ref, obj, imports);
    }
    if (obj instanceof Service) {
        return traverseService(store, ref, obj, imports);
    }
    if (obj instanceof Field) {
        return traverseField(store, ref, obj, imports);
    }
    if (obj instanceof Root) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, ref, obj.nested[key], imports, exports, isNested);
                return m;
            }, {
                type: 'Root',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    if (obj instanceof Namespace) {
        if (obj.nested) {
            return Object.keys(obj.nested).reduce((m, key) => {
                m.nested[key] = recursiveTraversal(store, ref, obj.nested[key], imports, exports, isNested);
                return m;
            }, {
                type: 'Namespace',
                nested: {}
            });
        } else {
            throw new Error('recursiveTraversal() cannot find protobufjs Type')
        }
    }
    throw new Error('recursiveTraversal() cannot find protobufjs Type')
};
