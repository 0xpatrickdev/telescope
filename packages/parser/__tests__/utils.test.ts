import cases from 'jest-in-case'
import { isRefIncluded } from '../src';
import { getTestProtoStore } from '../test-utils';

const store = getTestProtoStore();

cases(`isRefIncluded`, opts => {
    const ref = store.findProto(opts.value);
    const included = isRefIncluded(ref, {
        patterns: [
            'osmosis/**/gamm/**/query.proto'
        ],
        protos: [
            'akash/cert/v1beta2/query.proto'
        ],
        packages: [
            'cosmos.bank.v1beta1',
            'cosmos.gov.v1beta1',
            'evmos.erc20.v1'
        ]
    });
    expect(included).toMatchSnapshot();
},
    [
        'google/api/expr/conformance/v1alpha1/conformance_service.proto',
        'osmosis/gamm/v2/query.proto',
        'osmosis/gamm/v1beta1/query.proto',
        'osmosis/gamm/v1beta1/tx.proto'
    ].map(el => {
        return {
            name: el,
            value: el
        }
    })
);

const protos = [
    'cosmos/bank/v1beta1/bank.proto',
    'cosmos/bank/v1beta1/query.proto',
    'cosmos/bank/v1beta1/tx.proto',
    'cosmos/gov/v1beta1/gov.proto',
    'cosmos/gov/v1beta1/query.proto',
    'cosmos/gov/v1beta1/tx.proto',
    'google/api/expr/conformance/v1alpha1/conformance_service.proto',
    'osmosis/gamm/v1beta1/query.proto',
    'osmosis/gamm/v1beta1/tx.proto',
    'osmosis/gamm/v2/query.proto',
];
const protoNameValPairs = protos.map(el => {
    return {
        name: el,
        value: el
    }
});
cases(`isRefIncluded w minimatch`, opts => {
    const ref = store.findProto(opts.value);
    const included = isRefIncluded(ref, {
        protos: [
            'osmosis/**/gamm/**/query.proto'
        ],
        packages: [
            '*.gov.*',
            'evmos.erc20.*'
        ]
    });
    expect(included).toMatchSnapshot();
},
    protoNameValPairs
);

cases(`isRefIncluded w minimatch`, opts => {
    const ref = store.findProto(opts.value);
    const included = isRefIncluded(ref, {
        protos: [
            'osmosis/**/gamm/v1beta1/query.proto'
        ],
        packages: [
            'cosmos.*'
        ]
    });
    expect(included).toMatchSnapshot();
},
    protoNameValPairs
);

cases(`empty`, opts => {
    const ref = store.findProto(opts.value);
    const included = isRefIncluded(ref, {
        patterns: [],
        protos: [],
        packages: []
    });
    expect(included).toMatchSnapshot();
},
    protoNameValPairs
);

cases(`pkg`, opts => {
    const ref = store.findProto(opts.value);
    const included = isRefIncluded(ref, {
        patterns: [],
        protos: [],
        packages: ['osmosis.gamm.v1beta1']
    });
    expect(included).toMatchSnapshot();
},
    [
        'google/api/expr/conformance/v1alpha1/conformance_service.proto',
        'osmosis/gamm/v2/query.proto',
        'osmosis/gamm/v1beta1/query.proto',
        'osmosis/gamm/v1beta1/tx.proto'
    ].map(el => {
        return {
            name: el,
            value: el
        }
    })
);
