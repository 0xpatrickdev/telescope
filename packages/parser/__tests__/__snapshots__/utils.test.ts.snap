// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty cosmos/bank/v1beta1/bank.proto 1`] = `true`;

exports[`empty cosmos/bank/v1beta1/query.proto 1`] = `true`;

exports[`empty cosmos/bank/v1beta1/tx.proto 1`] = `true`;

exports[`empty cosmos/gov/v1beta1/gov.proto 1`] = `true`;

exports[`empty cosmos/gov/v1beta1/query.proto 1`] = `true`;

exports[`empty cosmos/gov/v1beta1/tx.proto 1`] = `true`;

exports[`empty google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `true`;

exports[`empty osmosis/gamm/v1beta1/query.proto 1`] = `true`;

exports[`empty osmosis/gamm/v1beta1/tx.proto 1`] = `true`;

exports[`empty osmosis/gamm/v2/query.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/bank/v1beta1/bank.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/bank/v1beta1/query.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/bank/v1beta1/tx.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/gov/v1beta1/gov.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/gov/v1beta1/query.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* cosmos/gov/v1beta1/tx.proto 1`] = `true`;

exports[`isRefIncluded cosmos.* google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`isRefIncluded cosmos.* osmosis/gamm/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded cosmos.* osmosis/gamm/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded cosmos.* osmosis/gamm/v2/query.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* cosmos/bank/v1beta1/bank.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* cosmos/bank/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* cosmos/bank/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* cosmos/gov/v1beta1/gov.proto 1`] = `true`;

exports[`isRefIncluded pkg *.gov.* cosmos/gov/v1beta1/query.proto 1`] = `true`;

exports[`isRefIncluded pkg *.gov.* cosmos/gov/v1beta1/tx.proto 1`] = `true`;

exports[`isRefIncluded pkg *.gov.* google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* osmosis/gamm/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* osmosis/gamm/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded pkg *.gov.* osmosis/gamm/v2/query.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/bank/v1beta1/bank.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/bank/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/bank/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/gov/v1beta1/gov.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/gov/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* cosmos/gov/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* osmosis/gamm/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* osmosis/gamm/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded pkg evmos.erc20.* osmosis/gamm/v2/query.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/bank/v1beta1/bank.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/bank/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/bank/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/gov/v1beta1/gov.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/gov/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto cosmos/gov/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto osmosis/gamm/v1beta1/query.proto 1`] = `true`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto osmosis/gamm/v1beta1/tx.proto 1`] = `true`;

exports[`isRefIncluded proto osmosis/**/gamm/**/*.proto osmosis/gamm/v2/query.proto 1`] = `true`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/bank/v1beta1/bank.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/bank/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/bank/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/gov/v1beta1/gov.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/gov/v1beta1/query.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto cosmos/gov/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto osmosis/gamm/v1beta1/query.proto 1`] = `true`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto osmosis/gamm/v1beta1/tx.proto 1`] = `false`;

exports[`isRefIncluded proto osmosis/**/gamm/**/query.proto osmosis/gamm/v2/query.proto 1`] = `true`;

exports[`pkg osmosis.gamm.v1beta1 google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`pkg osmosis.gamm.v1beta1 osmosis/gamm/v1beta1/query.proto 1`] = `true`;

exports[`pkg osmosis.gamm.v1beta1 osmosis/gamm/v1beta1/tx.proto 1`] = `true`;

exports[`pkg osmosis.gamm.v1beta1 osmosis/gamm/v2/query.proto 1`] = `false`;

exports[`proto akash/cert/v1beta2/query.proto akash/cert/v1beta2/query.proto 1`] = `true`;

exports[`proto akash/cert/v1beta2/query.proto google/api/expr/conformance/v1alpha1/conformance_service.proto 1`] = `false`;

exports[`proto akash/cert/v1beta2/query.proto osmosis/gamm/v1beta1/query.proto 1`] = `false`;

exports[`proto akash/cert/v1beta2/query.proto osmosis/gamm/v1beta1/tx.proto 1`] = `false`;

exports[`proto akash/cert/v1beta2/query.proto osmosis/gamm/v2/query.proto 1`] = `false`;
